dotnet/src/Connectors/Connectors.Memory.Pinecone/PineconeUtils.cs

```diff
@@ -0,0 +1,294 @@
+    /// <remarks>
+    ///  If the metadata size is too large, the document will be split into multiple documents.
+    ///  The text field will be removed from the metadata and the document ID will be set to {document ID}-{split counter}
+    ///  The text field will be added to the metadata of the split documents. The split documents will be returned in the
+    ///  same order as the original document.
+    /// </remarks>
+    internal static async IAsyncEnumerable<PineconeDocument> EnsureValidMetadataAsync(
+        IAsyncEnumerable<PineconeDocument> documents)
+    {
+        await foreach (PineconeDocument document in documents)
+        {
+            if (document.Metadata == null || GetMetadataSize(document.Metadata) <= MaxMetadataSize)
+            {
+                yield return document;
+
+                continue;
+            }
+
+            if (!document.Metadata.TryGetValue("text", out object value))
+            {
+                yield return document;
+
+                continue;
+            }
+
+            Console.WriteLine($"Splitting document {document.Id} into multiple documents because metadata size is too large");
+
+            string text = value as string ?? string.Empty;
+            int textSize = Encoding.UTF8.GetByteCount(text);
+            document.Metadata.Remove("text");
+            int remainingMetadataSize = GetMetadataSize(document.Metadata);
+
+            int splitCounter = 0;
+            int textIndex = 0;
+
+            while (textSize > 0)
+            {
+                int availableSpace = MaxMetadataSize - remainingMetadataSize;
+                int textSplitSize = Math.Min(textSize, availableSpace);
+
+                while (textSplitSize > 0 && Encoding.UTF8.GetByteCount(text.ToCharArray(textIndex, textSplitSize)) > availableSpace)
+                {
+                    textSplitSize--;
+                }
+
+                string splitText = text.Substring(textIndex, textSplitSize);
+                textIndex += textSplitSize;
+                textSize -= Encoding.UTF8.GetByteCount(splitText);
+
+                PineconeDocument splitDocument = PineconeDocument.Create($"{document.Id}_{splitCounter}", document.Values)
+                    .WithMetadata(new Dictionary<string, object>(document.Metadata))
+                    .WithSparseValues(document.SparseValues);
+                splitDocument.Metadata!["text"] = splitText;
+
+                yield return splitDocument;
+
+                splitCounter++;
+            }
+
+            Console.WriteLine($"Split document {document.Id} into {splitCounter} documents");
+        }
+    }
+
+    /// <summary>
+    ///  Utility method to split a stream of documents into batches of a given size.
+    /// </summary>
+    /// <param name="data"></param>
+    /// <param name="batchSize"></param>
+    /// <returns></returns>
+    /// <remarks>
+    ///  The recommended batch size as per Pinecone documentation is 100.
```

Comment:
Let's leave here link to the Pinecone documentation about recommended batch size.